### PR TITLE
[bitnami/harbor]Add harbor exporter

### LINKED ISSUE #5784: [bitnami/harbor] metric exporter is missing
**Which chart**:
Harbor 9.7.x
**Describe the bug**
The metrics exporter added with appVersion 2.2.0 is missing
**To Reproduce**

install harbor, notice no metgric exporter
**Expected behavior**
The metric exporter is installed.

**Additional context**
metric exporter in officvial chart: https://github.com/goharbor/harbor-helm/tree/master/templates/exporter
release notes adding the exporter: https://github.com/goharbor/harbor/releases/tag/v2.2.0

### PR DESCRIPTION
**Pull request by fmulero**
**Description of the change**
Include harbor exporter component and enable metrics.

**Benefits**
Observability and Monitoring

**Possible drawbacks**
Not identified

**Applicable issues**

  - fixes #5784 

**Additional information**

**Checklist**

- [X] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [X] Variables are documented in the values.yaml and added to the README.md using [readme-generator-for-helm](https://github.com/bitnami-labs/readme-generator-for-helm)
- [X] Title of the PR starts with chart name (e.g. [bitnami/<name_of_the_chart>])
- [X] All commits signed off and in agreement of [Developer Certificate of Origin (DCO)](https://github.com/bitnami/charts/blob/master/CONTRIBUTING.md#sign-your-work)

### CODE REVIEW

File: bitnami/harbor/templates/_helpers.tpl [hunk 1]
```diff
@@ -439,6 +443,13 @@ Return the proper Harbor Core image name
 {{- include "common.images.image" ( dict "imageRoot" .Values.core.image "global" .Values.global ) -}}
 {{- end -}}
 
+{{/*
+Return the proper Harbor Exporter image name
+*/}}
+{{- define "harbor.exporter.image" -}}
+{{- include "common.images.image" ( dict "imageRoot" .Values.exporter.image "global" .Values.global ) -}}
+{{- end -}}
```
Comment (juan131) [inline bitnami/harbor/templates/_helpers.tpl:451]: Nice! Please also update "harbor.imagePullSecrets" (line 541) including this image in the list.

File: bitnami/harbor/templates/core/core-cm-envvars.yaml [hunk 2]
```diff
@@ -52,3 +52,10 @@ data:
   HTTPS_PROXY: {{ .Values.proxy.httpsProxy | quote }}
   NO_PROXY: {{ include "harbor.noProxy" . | quote }}
   {{- end }}
+  {{- if .Values.metrics.enabled}}
+  METRIC_ENABLE: "true"
+  METRIC_PATH: "{{ .Values.metrics.path }}"
+  METRIC_PORT: "{{ .Values.core.containerPorts.metrics }}"
```
Comment (juan131) [inline bitnami/harbor/templates/core/core-cm-envvars.yaml:58]: Use quote instead:
```suggestion
  METRIC_PATH: {{ .Values.metrics.path | quote }}
  METRIC_PORT: {{ .Values.core.containerPorts.metrics | quote }}
```

File: bitnami/harbor/templates/core/core-svc.yaml [hunk 3]
```diff
@@ -19,5 +19,10 @@ spec:
     - name: {{ ternary "https" "http" .Values.internalTLS.enabled }}
       port: {{ ternary .Values.core.service.ports.https .Values.core.service.ports.http .Values.internalTLS.enabled }}
       targetPort: {{ ternary "https" "http" .Values.internalTLS.enabled }}
+    {{- if .Values.metrics.enabled }}
+    - name: {{ template "harbor.metricsPortName" . }}
+      port: {{ .Values.core.service.ports.metrics }}
+      targetPort: {{ template "harbor.metricsPortName" . }}
```
Comment (juan131) [inline bitnami/harbor/templates/core/core-svc.yaml:25]: Again.. I don't see any value on using a helper that hardcodes the name. Also, add in the name the protocol as prefix:
```suggestion
    - name: http-metrics
      port: {{ .Values.core.service.ports.metrics }}
      targetPort: metrics
```

File: bitnami/harbor/templates/exporter/exporter-cm-envvars.yaml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  annotations: {{- include "common.tplvalues.render" ( dict "value" .Values.commonAnnotations "context" $ ) | nindent 4 }}
+  {{- end }}
+data:
+  LOG_LEVEL: {{ .Values.logLevel | quote }}
+  HARBOR_EXPORTER_PORT: {{ .Values.exporter.containerPorts.metrics | quote }}
+  HARBOR_EXPORTER_METRICS_PATH: {{ .Values.metrics.path | quote }}
+  HARBOR_METRIC_NAMESPACE: harbor
+  HARBOR_METRIC_SUBSYSTEM: exporter
+  HARBOR_REDIS_URL: {{ include "harbor.redisForCore" . | quote }}
+  HARBOR_REDIS_NAMESPACE: harbor_job_service_namespace
```
Comment (juan131) [inline bitnami/harbor/templates/exporter/exporter-cm-envvars.yaml:21]: Hardcoded?

File: bitnami/harbor/templates/_helpers.tpl [hunk 5]
```diff
@@ -530,6 +541,14 @@ Return the proper Docker Image Registry Secret Names
 {{- include "common.images.pullSecrets" (dict "images" (list .Values.core.image .Values.portal.image .Values.jobservice.image .Values.clair.server.image .Values.clair.adapter.image .Values.chartmuseum.image .Values.trivy.image .Values.notary.server.image .Values.notary.signer.image .Values.registry.server.image .Values.registry.controller.image .Values.nginx.image .Values.volumePermissions.image) "global" .Values.global) -}}
 {{- end -}}
 
+{{/*
+Return the metrics port name
+*/}}
+{{- define "harbor.metricsPortName" -}}
+{{- printf "metrics" -}}
+{{- end -}}
```
Comment (juan131) [inline bitnami/harbor/templates/_helpers.tpl:549]: I don't find this helper really useful. It's defining a hardcoded "metrics" values, therefore I'll remove it.

File: bitnami/harbor/templates/exporter/exporter-dpl.yaml [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+                  key: HARBOR_ADMIN_PASSWORD
+            {{- if .Values.exporter.extraEnvVars }}
+            {{- include "common.tplvalues.render" (dict "value" .Values.exporter.extraEnvVars "context" $) | nindent 12 }}
+            {{- end }}
+          envFrom:
+            - configMapRef:
+                name: "{{ include "harbor.exporter" . }}-envvars"
+          ports:
+            - containerPort: {{ .Values.exporter.containerPorts.metrics }}
+              name: {{ include "harbor.metricsPortName" . }}
```
Comment (juan131) [inline bitnami/harbor/templates/exporter/exporter-dpl.yaml:118]: Again, hardcode the name:
```suggestion
              name: metrics
```

File: bitnami/harbor/templates/exporter/exporter-dpl.yaml [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+          {{- if not .Values.diagnosticMode.enabled }}
+          {{- if .Values.exporter.startupProbe.enabled }}
+          startupProbe: {{- include "common.tplvalues.render" (dict "value" (omit .Values.exporter.startupProbe "enabled") "context" $) | nindent 12 }}
+            httpGet:
+              path: {{ .Values.metrics.path | quote }}
+              port: {{ include "harbor.metricsPortName" . }}
```
Comment (juan131) [inline bitnami/harbor/templates/exporter/exporter-dpl.yaml:124]:
```suggestion
              port: metrics
```

File: bitnami/harbor/templates/exporter/exporter-dpl.yaml [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+              # Metrics are exposed only though HTTP https://github.com/goharbor/harbor/issues/16252
+              scheme: HTTP
+          {{- else if .Values.exporter.customStartupProbe }}
+          startupProbe: {{- include "common.tplvalues.render" (dict "value" .Values.exporter.customStartupProbe "context" $) | nindent 12 }}
+          {{- end }}
+          {{- if .Values.exporter.livenessProbe.enabled }}
+          livenessProbe: {{- include "common.tplvalues.render" (dict "value" (omit .Values.exporter.livenessProbe "enabled") "context" $) | nindent 12 }}
+            httpGet:
+              path: {{ .Values.metrics.path | quote }}
+              port: {{ include "harbor.metricsPortName" . }}
```
Comment (juan131) [inline bitnami/harbor/templates/exporter/exporter-dpl.yaml:134]:
```suggestion
              port: metrics
```

File: bitnami/harbor/templates/exporter/exporter-dpl.yaml [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+              scheme: HTTP
+          {{- else if .Values.exporter.customLivenessProbe }}
+          livenessProbe: {{- include "common.tplvalues.render" (dict "value" .Values.exporter.customLivenessProbe "context" $) | nindent 12 }}
+          {{- end }}
+          {{- if .Values.exporter.readinessProbe.enabled }}
+          readinessProbe: {{- include "common.tplvalues.render" (dict "value" (omit .Values.exporter.readinessProbe "enabled") "context" $) | nindent 12 }}
+            httpGet:
+              path: {{ .Values.metrics.path | quote }}
+              port: {{ include "harbor.metricsPortName" . }}
```
Comment (juan131) [inline bitnami/harbor/templates/exporter/exporter-dpl.yaml:143]:
```suggestion
              port: metrics
```

File: bitnami/harbor/templates/exporter/exporter-dpl.yaml [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+          lifecycle: {{- include "common.tplvalues.render" (dict "value" .Values.exporter.lifecycleHooks "context" $) | nindent 12 }}
+          {{- end }}
+          {{- if .Values.exporter.resources }}
+          resources: {{- toYaml .Values.exporter.resources | nindent 12 }}
+          {{- end }}
+          volumeMounts:
+            {{- if .Values.internalTLS.enabled }}
+            - name: internal-tls-certs
+              mountPath: /etc/harbor/ssl/exporter
+            {{- end }}
```
Comment (juan131) [inline bitnami/harbor/templates/exporter/exporter-dpl.yaml:159]: is the exporter really taken into account these certs?

File: bitnami/harbor/templates/jobservice/jobservice-dpl.yaml [hunk 11]
```diff
@@ -158,6 +158,10 @@ spec:
           ports:
             - containerPort: {{ ternary .Values.jobservice.containerPorts.https .Values.jobservice.containerPorts.http .Values.internalTLS.enabled }}
               name: {{ ternary "https" "http" .Values.internalTLS.enabled }}
+            {{- if .Values.metrics.enabled }}
+            - containerPort: {{ .Values.jobservice.containerPorts.metrics }}
+              name: {{ include "harbor.metricsPortName" . }}
```
Comment (juan131) [inline bitnami/harbor/templates/jobservice/jobservice-dpl.yaml:163]:
```suggestion
              name: metrics
```

File: bitnami/harbor/templates/jobservice/jobservice-svc.yaml [hunk 12]
```diff
@@ -16,5 +16,10 @@ spec:
     - name: {{ ternary "https" "http" .Values.internalTLS.enabled }}
       port: {{ ternary .Values.jobservice.service.ports.https .Values.jobservice.service.ports.http .Values.internalTLS.enabled }}
       targetPort: {{ ternary "https" "http" .Values.internalTLS.enabled }}
+    {{- if .Values.metrics.enabled }}
+    - name: {{ template "harbor.metricsPortName" . }}
+      port: {{ .Values.jobservice.service.ports.metrics }}
+      targetPort: {{ template "harbor.metricsPortName" . }}
```
Comment (juan131) [inline bitnami/harbor/templates/jobservice/jobservice-svc.yaml:22]: Same here:
```suggestion
    - name: http-metrics
      port: {{ .Values.jobservice.service.ports.metrics }}
      targetPort: metrics
```

File: bitnami/harbor/templates/registry/registry-svc.yaml [hunk 13]
```diff
@@ -16,5 +16,10 @@ spec:
       port: {{ ternary .Values.registry.server.service.ports.https .Values.registry.server.service.ports.http .Values.internalTLS.enabled }}
     - name: controller
       port: {{ ternary .Values.registry.controller.service.ports.https .Values.registry.controller.service.ports.http .Values.internalTLS.enabled }}
+    {{- if .Values.metrics.enabled }}
+    - name: {{ template "harbor.metricsPortName" . }}
+      port: {{ .Values.registry.server.service.ports.metrics }}
+      targetPort: {{ template "harbor.metricsPortName" . }}
```
Comment (juan131) [inline bitnami/harbor/templates/registry/registry-svc.yaml:22]: Same here:
```suggestion
    - name: http-metrics
      port: {{ .Values.registry.server.service.ports.metrics }}
      targetPort: metrics
```

File: bitnami/harbor/templates/registry/registry-dpl.yaml [hunk 14]
```diff
@@ -155,6 +155,10 @@ spec:
               name: registry
             - containerPort: {{ .Values.registry.server.containerPorts.debug }}
               name: debug
+            {{- if .Values.metrics.enabled }}
+            - containerPort: {{ .Values.registry.server.containerPorts.metrics }}
+              name: {{ include "harbor.metricsPortName" . }}
```
Comment (juan131) [inline bitnami/harbor/templates/registry/registry-dpl.yaml:160]:
```suggestion
              name: metrics
```

**Review by juan131 — CHANGES_REQUESTED** [review]
Thanks for your PR @fmulero ! Please check my comments when you have a chance

File: bitnami/harbor/templates/exporter/exporter-cm-envvars.yaml (hunk 4 above)
Comment (fmulero) [reply bitnami/harbor/templates/exporter/exporter-cm-envvars.yaml:21]: Good catch!

File: bitnami/harbor/templates/exporter/exporter-dpl.yaml (hunk 10 above)
Comment (fmulero) [reply bitnami/harbor/templates/exporter/exporter-dpl.yaml:159]: For internal connections to harbor-core. Only CA.crt certificate is needed

File: bitnami/harbor/templates/jobservice/jobservice-cm.yaml [hunk 15]
```diff
@@ -25,8 +25,8 @@ data:
       workers: {{ .Values.jobservice.maxJobWorkers }}
       backend: "redis"
       redis_pool:
-        redis_url: "{{ template "harbor.redisForJobservice" . }}"
-        namespace: "harbor_job_service_namespace"
+        redis_url: "{{ include "harbor.redisForJobservice" . }}"
```
Comment (juan131) [inline bitnami/harbor/templates/jobservice/jobservice-cm.yaml:28]: Use "quote" instead:
```suggestion
        redis_url: {{ include "harbor.redisForJobservice" . | quote }}
```

File: bitnami/harbor/values.yaml [hunk 16]
```diff
@@ -3701,6 +3716,251 @@ trivy:
...
+  ##
+  extraEnvVarsSecret: ""
+  ## @param exporter.containerPorts.metrics Harbor Exporter HTTP container port
+  ##
+  containerPorts:
+    metrics: 8001
+  ## @param exporter.replicaCount The replica count
+  ##
+  replicaCount: 1
+    ## Harbor Exporter containers' liveness probe
```
Comment (juan131) [inline bitnami/harbor/values.yaml:3777]: Wrong indentation:
```suggestion
  ## Harbor Exporter containers' liveness probe
```

File: bitnami/harbor/values.yaml [hunk 17]
```diff
@@ -3701,6 +3716,251 @@ trivy:
...
+  ## @param exporter.livenessProbe.enabled Enable livenessProbe
+  ## @param exporter.livenessProbe.initialDelaySeconds Initial delay seconds for livenessProbe
+  ## @param exporter.livenessProbe.periodSeconds Period seconds for livenessProbe
+  ## @param exporter.livenessProbe.timeoutSeconds Timeout seconds for livenessProbe
+  ## @param exporter.livenessProbe.failureThreshold Failure threshold for livenessProbe
+  ## @param exporter.livenessProbe.successThreshold Success threshold for livenessProbe
+  ##
+  livenessProbe:
+    enabled: true
+    initialDelaySeconds: 20
```
Comment (juan131) [inline bitnami/harbor/values.yaml:3788]: 20 seconds seems like a very high value for "initialDelaySeconds". Why did you choose this value?

File: bitnami/harbor/values.yaml [hunk 18]
```diff
@@ -3701,6 +3716,251 @@ trivy:
...
+    ## @param exporter.nodeAffinityPreset.values Harbor Exporter Node label values to match. Ignored if `affinity` is set.
+    ## E.g.
+    ## values:
+    ##   - e2e-az1
+    ##   - e2e-az2
+    ##
+    values: []
+  ## @param exporter.affinity Harbor Exporter Affinity for pod assignment
+  ## Ref: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#affinity-and-anti-affinity
+  ## Note: podAffinityPreset, podAntiAffinityPreset, and  nodeAffinityPreset will be ignored when it's set
```
Comment (juan131) [inline bitnami/harbor/values.yaml:3922]:
```suggestion
  ## Note: `exporter.podAffinityPreset`, `exporter.podAntiAffinityPreset`, and `exporter.nodeAffinityPreset` will be ignored when it's set
```

File: bitnami/harbor/values.yaml [hunk 19]
```diff
@@ -3701,6 +3716,251 @@ trivy:
...
+  podAffinityPreset: ""
+  ## @param exporter.podAntiAffinityPreset Harbor Exporter Pod anti-affinity preset. Ignored if `affinity` is set. Allowed values: `soft` or `hard`
+  ## Ref: https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity
+  ##
+  podAntiAffinityPreset: soft
+  ## Node affinity preset
+  ## Ref: https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#node-affinity
+  ##
+  nodeAffinityPreset:
+    ## @param exporter.nodeAffinityPreset.type Harbor Exporter Node affinity preset type. Ignored if `affinity` is set. Allowed values: `soft` or `hard`
```
Comment (juan131) [inline bitnami/harbor/values.yaml:3905]:
```suggestion
    ## @param exporter.nodeAffinityPreset.type Harbor Exporter Node affinity preset type. Ignored if `exporter.affinity` is set. Allowed values: `soft` or `hard`
```

File: bitnami/harbor/values.yaml [hunk 20]
```diff
@@ -3701,6 +3716,251 @@ trivy:
...
+    ##
+    type: ""
+    ## @param exporter.nodeAffinityPreset.key Harbor Exporter Node label key to match Ignored if `affinity` is set.
```
Comment (juan131) [inline bitnami/harbor/values.yaml:3908]:
```suggestion
    ## @param exporter.nodeAffinityPreset.key Harbor Exporter Node label key to match Ignored if `exporter.affinity` is set.
```

File: bitnami/harbor/values.yaml [hunk 21]
```diff
@@ -3701,6 +3716,251 @@ trivy:
...
+    ## E.g.
+    ## key: "kubernetes.io/e2e-az-name"
+    ##
+    key: ""
+    ## @param exporter.nodeAffinityPreset.values Harbor Exporter Node label values to match. Ignored if `affinity` is set.
```
Comment (juan131) [inline bitnami/harbor/values.yaml:3913]:
```suggestion
    ## @param exporter.nodeAffinityPreset.values Harbor Exporter Node label values to match. Ignored if `exporter.affinity` is set.
```

File: bitnami/harbor/values.yaml [hunk 22]
```diff
@@ -3701,6 +3716,251 @@ trivy:
...
+    ## Example:
+    ## limits:
+    ##    cpu: 500m
+    ##    memory: 1Gi
+    limits: {}
+    ## Examples:
+    ## requests:
+    ##    cpu: 250m
+    ##    memory: 256Mi
+    requests: {}
```
Comment (juan131) [inline bitnami/harbor/values.yaml:3851]: To be aligned with the other components:
```suggestion
  ## Harbor Exporter resource requests and limits
  ## ref: https://kubernetes.io/docs/user-guide/compute-resources/
  ## @param exporter.resources.limits The resources limits for the Harbor Exporter containers
  ## @param exporter.resources.requests The requested resources for the Harbor Exporter containers
  ##
  resources:
    limits: {}
    requests: {}
```

File: bitnami/harbor/values.yaml [hunk 23]
```diff
@@ -3701,6 +3716,251 @@ trivy:
...
+    ## ref: https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/
+    ## e.g:
+    ## pullSecrets:
+    ##   - myRegistryKeySecretName
+    ##
+    pullSecrets: []
+    ## Set to true if you would like to see extra information on logs
+    ## It turns BASH and/or NAMI debugging in the image
+    ##
+    debug: false
```
Comment (juan131) [inline bitnami/harbor/values.yaml:3751]: Nami doesn't exist anymore:
```suggestion
    ## Enable debug mode
    ##
    debug: false
```

**Review by juan131 — CHANGES_REQUESTED** [review]
Some minor requests

File: bitnami/harbor/values.yaml (hunk 17 above)
Comment (fmulero) [reply bitnami/harbor/values.yaml:3788]: In the upstream project the `intialDelaySeconds` was 300. I reduced and tested these values and they works fine. I've noticed that some times the  exporter spin up before the `jobservice` and in that case the set of metrics returned by the exporter is not complete for some seconds. Maybe is worth to include an init container to wait for the `jobservice`

File: bitnami/harbor/templates/jobservice/jobservice-cm.yaml (hunk 15 above)
Comment (fmulero) [reply bitnami/harbor/templates/jobservice/jobservice-cm.yaml]: Done

File: bitnami/harbor/values.yaml (hunk 17 above)
Comment (juan131) [reply bitnami/harbor/values.yaml:3787]: Fair enough

File: bitnami/harbor/templates/metrics/metrics-svcmonitor.yaml [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+    app.kubernetes.io/component: metrics
+  {{- if .Values.commonAnnotations }}
+  annotations: {{- include "common.tplvalues.render" (dict "value" .Values.commonAnnotations "context" $) | nindent 4 }}
+  {{- end }}
+spec:
+  {{- if .Values.metrics.serviceMonitor.jobLabel }}
+  jobLabel: {{ .Values.metrics.serviceMonitor.jobLabel }}
+  {{- end }}
+  selector:
+    matchLabels: {{- include "common.labels.matchLabels" . | nindent 6 }}
```
Comment (migruiz4) [inline bitnami/harbor/templates/metrics/metrics-svcmonitor.yaml:23]: Could you please add support for `metrics.serviceMonitor.selector`? It exists at `values.yaml` but is not implemented.
```suggestion
  selector:
    matchLabels: {{- include "common.labels.matchLabels" . | nindent 6 }}
      {{- if .Values.metrics.serviceMonitor.selector }}
      {{- include "common.tplvalues.render" (dict "value" .Values.metrics.serviceMonitor.selector "context" $) | nindent 6 }}
      {{- end }}
```

**Review by migruiz4 — APPROVED** [review]
Looks good to me, but please check below comment

**Review by juan131 — APPROVED** [review]

File: bitnami/harbor/templates/metrics/metrics-svcmonitor.yaml (hunk 24 above)
Comment (fmulero) [reply bitnami/harbor/templates/metrics/metrics-svcmonitor.yaml:23]: Good Catch!!

**Review by migruiz4 — DISMISSED** [review]

**Comment by fmulero** [comment]
Closed to avoid more noise due to a rebase issue.  The follow up is #9377

**Pull request closed without merging**